### PR TITLE
slow_conv3d: Avoid dispatch in parallel region

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66165
* #65759
* #65758
* #65757
* __->__ #65737
* #65726

See gh-56794

Avoid dispatch inside of parallel_for by:
- Replacing Tensor slicing with TensorAccessor
- Copy bias into output only once, outside of the parallel region
- Replaces `addmm_` and `baddbmm_` with direct calls to gemm.

Differential Revision: [D31257874](https://our.internmc.facebook.com/intern/diff/D31257874)

### CODE REVIEW

**Comment by ngimel** [comment]
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31257874).

**Comment by ngimel** [comment]
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31257874).

**Review by ngimel — APPROVED** [review]

**Comment by ngimel** [comment]
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31257874).

File: aten/src/ATen/native/mkl/LinearAlgebra.cpp [hunk 1]
```diff
@@ -100,6 +133,63 @@ static inline void gemm_batched(const CBLAS_TRANSPOSE trans_A, const CBLAS_TRANS
...
+static CBLAS_TRANSPOSE to_cblas(TransposeType x) {
+  switch (x) {
+    case TransposeType::NoTranspose: return CblasNoTrans;
+    case TransposeType::Transpose: return CblasTrans;
+    case TransposeType::ConjTranspose: return CblasConjTrans;
+  }
+  TORCH_INTERNAL_ASSERT(false, "Unknown TransposeType");
+}
+
+void mkl_gemm_batched(
```
Comment (ngimel) [inline aten/src/ATen/native/mkl/LinearAlgebra.cpp:145]: it seems these functions are the same as `gemm_batched` above, modulo type of trans_A/trans_B, why do we need 2 copies?

File: aten/src/ATen/native/CPUBlas.cpp [hunk 2]
```diff
@@ -251,6 +255,154 @@ void gemm(
...
+void gemm_batched(
+    TransposeType transa, TransposeType transb,
+    int64_t batch_size, int64_t m, int64_t n, int64_t k,
+    scalar_t alpha,
+    const scalar_t **a, int64_t lda,
+    const scalar_t **b, int64_t ldb,
+    scalar_t beta,
+    scalar_t **c, int64_t ldc) {
+  if (batch_size == 1) {
+    return gemm(transa, transb, m, n, k, alpha, a[0], lda, b[0], ldb, beta, c[0], ldc);
```
Comment (ngimel) [inline aten/src/ATen/native/CPUBlas.cpp:306]: this is nice! I was wondering if calling gemm_batched regardles of `groups` is going to be ok, but with this special case it is

**Review by ngimel — COMMENTED** [review]
Thanks! Adding batched gemm to at::native::cpublas is great, I have a small question about code duplication

File: aten/src/ATen/native/mkl/LinearAlgebra.cpp (hunk 1 above)
Comment (peterbell10) [reply aten/src/ATen/native/mkl/LinearAlgebra.cpp:145]: The biggest difference is `ColMajor` vs. `RowMajor`. In a follow up I can remove `_baddbmm_mkl_` entirely though. Just have the normal version call `cpublas::gemm_batched_with_stride`.

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**